feat: Add Legal Agent Bench resource server (#1976)
NVIDIA-NeMo/Gym: resources_servers/legal_agent_bench/README.md

judge_base_url: https://your-judge-endpoint.example/v1
judge_api_key: your-judge-key
judge_model_name: your-judge-model
```

The judge credentials are injected only into the regenerated, gitignored
runtime task tree. They are never written into the pristine cache.

## Prepare explicitly (recommended)

From the repository root:

```bash
python resources_servers/legal_agent_bench/prepare.py
```

The command downloads the pinned LAB source archive from GitHub with retries and visible
progress, verifies SHA-256
`e45cbdf3236b22866e034bcc62fb23bf00ef2f2e49db7a0cd8a4b07dbae9212c`,
rejects unsafe archive entries, generates deterministic Harbor tasks, and
builds each cache in staging before replacing the previous valid cache. A
handled preparation failure leaves the previous cache in place.

Useful options:

```bash
python resources_servers/legal_agent_bench/prepare.py --asset tasks
python resources_servers/legal_agent_bench/prepare.py --asset skills
python resources_servers/legal_agent_bench/prepare.py --force
python resources_servers/legal_agent_bench/prepare.py \
  --tasks-dir /custom/task-cache \
  --skills-dir /custom/skills-cache
```

## Prepare through the normal Gym workflow

> The committed `data/all.jsonl` is a lightweight 1,749-row task index, so

Reviewer: Actually I'm not sure we should commit it. Does it require maintenance? If so, is it documented anywhere how to re-fresh this file if it gets outdated? I think having commands/scripts for creating it would be a better strategy.

Author: Removed all.jsonl and set it up to download as part of prepare.py

> `ng_prepare_data` works without downloading binary assets. If you skip

Reviewer: Let's use new CLI commands (see [migration guide](https://docs.nvidia.com/nemo/gym/reference/cli-commands#migrating-from-the-legacy-commands))
```suggestion
`gym dataset collate` works without downloading binary assets. If you skip
```

> `prepare.py`, `ng_run` prepares missing task and skill caches during startup.

Reviewer: ```suggestion
`prepare.py`, `gym env start` prepares missing task and skill caches during startup.
```

Author: Confirmed full flow works with updated `gym` commands and updated the README to show the appropriate commands

Set the two configs used by the benchmark:

```bash
CONFIG_PATHS=resources_servers/legal_agent_bench/configs/legal_agent_bench.yaml,responses_api_models/vllm_model/configs/vllm_model.yaml
```

Validate the five examples:

```bash

> ng_prepare_data "+config_paths=[${CONFIG_PATHS}]" \

Reviewer: With the new CLI this should be more or less:
```
gym dataset collate \
  --resources-server legal_agent_bench \
  --model-type vllm_model \
  --output-dir results/legal_agent_bench_prepare \
  --mode example_validation
```

  +output_dirpath=results/legal_agent_bench_prepare \
  +mode=example_validation
```

To collate the full validation index instead:

```bash

> ng_prepare_data "+config_paths=[${CONFIG_PATHS}]" \

Reviewer: ```
gym dataset collate \
  --resources-server legal_agent_bench \
  --model-type vllm_model \
  --output-dir results/legal_agent_bench_prepare \
  --mode train_preparation
```

  +output_dirpath=results/legal_agent_bench_prepare \
  +mode=train_preparation
```

> ## Run and smoke test

Reviewer: It would be good to include info on running uni tests for this server somewhere in the README:

```
gym env test --resources-server legal_agent_bench
```

Author: Added a "Test the environment" section to the readme to address this

Start the servers:

```bash

> ng_run "+config_paths=[${CONFIG_PATHS}]"

Reviewer: ```
gym env start \
  --resources-server legal_agent_bench \
  --model-type vllm_model
```

```

On a clean cache, startup visibly downloads and prepares the pinned assets.
Every startup regenerates the runtime task tree so old judge credentials cannot
be reused.

In a second activated terminal, collect one rollout:

```bash
CONFIG_PATHS=resources_servers/legal_agent_bench/configs/legal_agent_bench.yaml,responses_api_models/vllm_model/configs/vllm_model.yaml

> ng_collect_rollouts "+config_paths=[${CONFIG_PATHS}]" \

Reviewer: ```
gym eval run --no-serve \
  --agent legal_agent_bench_harbor_agent \
  --input resources_servers/legal_agent_bench/data/example.jsonl \
  --output results/legal_agent_bench_smoke_rollout.jsonl \
  --concurrency 1 \
  --limit 1
```

  +agent_name=legal_agent_bench_harbor_agent \
  +input_jsonl_fpath=resources_servers/legal_agent_bench/data/example.jsonl \
  +output_jsonl_fpath=results/legal_agent_bench_smoke_rollout.jsonl \
  +num_samples_in_parallel=1 \
  +limit=1
```

The default `full_task` reward is LAB's official all-criteria score: a task
earns `1.0` only when every criterion passes. For diagnostic partial credit,
start with:

```bash

> ng_run "+config_paths=[${CONFIG_PATHS}]" \

Reviewer: ```
gym env start \
  --resources-server legal_agent_bench \
  --model-type vllm_model \
  +legal_agent_bench.resources_servers.legal_agent_bench.reward_mode=criteria_pass_rate
```

  +legal_agent_bench.resources_servers.legal_agent_bench.reward_mode=criteria_pass_rate
```

The verifier evaluates up to six criteria concurrently, matching the upstream LAB
default. Adjust the resource setting when the judge endpoint has a lower
concurrency limit:

```bash

> ng_run "+config_paths=[${CONFIG_PATHS}]" \

Reviewer: ```
gym env start \
  --resources-server legal_agent_bench \
  --model-type vllm_model \
  +legal_agent_bench.resources_servers.legal_agent_bench.judge_parallelism=2
```

  +legal_agent_bench.resources_servers.legal_agent_bench.judge_parallelism=2
```

This setting is forwarded to each task verifier as
`LAB_JUDGE_PARALLELISM`.

The task container requires network access because its verifier calls the
configured judge endpoint. The agent and verifier share that container, so the
agent also has network access during a rollout. This differs from the upstream LAB
closed-network reference sandbox and should be recorded when comparing runs.

## Caches and outputs

The default paths are:

- Pristine generated tasks: `data/cache/harbor_tasks/legal_agent_bench`
- Public skills: `data/cache/harness/skills`
- Credential-bearing runtime tasks: `data/runtime/harbor_tasks/legal_agent_bench`
- Harbor jobs: `results/legal_agent_bench/harbor_jobs`

> - Rollout output: the path passed to `ng_collect_rollouts`

Reviewer: ```suggestion
- Rollout output: the path passed to `gym eval run`
```
